autopush/tests/test_integration.py

```diff
@@ -1936,6 +1939,253 @@ def test_registration(self):
+        for profile in adm.profiles:
+            adm.profiles[profile]._request = self._mock_send
+
+    def test_bad_config(self):
+        from autopush.router.adm import ADMRouter
+        with pytest.raises(IOError):
+            ADMRouter(
+                self.ep.conf,
+                {
+                    "dev": {
+                        "app_id": "amzn1.application.StringOfStuff",
+                        "client_id":
+                            "amzn1.application-oa2-client.ev4nM0reStuff",
+                        "collapseKey": "simplepush",
+                    }
+                },
+                self.ep.db.metrics,
+            )
+
+    @inlineCallbacks
+    def test_missing_token_registration(self):
+        self._add_router()
+        url = "{}/v1/{}/{}/registration".format(
+            self.ep.conf.endpoint_url,
+            "adm",
+            "dev",
+        )
+        self._mock_reply.status_code = 200
+        self._mock_reply.json.return_value = {
+            "access_token": "token",
+            "expires_in": 3000
+        }
+        response, body = yield _agent("POST", url, body=json.dumps(
+            {"foo": self.token}
+        ))
+        assert response.code == 401
+        url = "{}/v1/{}/{}/registration".format(
+            self.ep.conf.endpoint_url,
+            "adm",
+            "foo",
+        )
+        response, body = yield _agent("POST", url, body=json.dumps(
+            {"token": self.token}
+        ))
+        assert response.code == 410
+
+    @inlineCallbacks
+    def test_successful(self):
+        self._add_router()
+        url = "{}/v1/{}/{}/registration".format(
+            self.ep.conf.endpoint_url,
+            "adm",
+            "dev",
+        )
+        self._mock_reply.status_code = 200
+        self._mock_reply.json.return_value = {
+            "access_token": "token",
+            "expires_in": 3000
+        }
+        response, body = yield _agent("POST", url, body=json.dumps(
+            {"token": self.token}
+        ))
+        assert response.code == 200
+        jbody = json.loads(body)
+
+        print("Response = {}", body)
+
+        data = ("\xa2\xa5\xbd\xda\x40\xdc\xd1\xa5\xf9\x6a\x60\xa8\x57\x7b\x48"
+                "\xe4\x43\x02\x5a\x72\xe0\x64\x69\xcd\x29\x6f\x65\x44\x53\x78"
+                "\xe1\xd9\xf6\x46\x26\xce\x69")
+        crypto_key = ("keyid=p256dh;dh=BAFJxCIaaWyb4JSkZopERL9MjXBeh3WdBxew"
+                      "SYP0cZWNMJaT7YNaJUiSqBuGUxfRj-9vpTPz5ANmUYq3-u-HWOI")
+        salt = "keyid=p256dh;salt=S82AseB7pAVBJ2143qtM3A"
+        content_encoding = "aesgcm"
+
+        self._mock_reply.json.return_value = dict(access_token="access.123")
+
+        response, body = yield _agent(
+            'POST',
+            str(jbody['endpoint']),
+            headers=Headers({
+                "crypto-key": [crypto_key],
+                "encryption": [salt],
+                "ttl": ["0"],
+                ""
```

Comment:
kill this stray line (it just str concats into "content-encoding")

Reply:
Heh, had that in there for readability but then realized that i was the only person that cared.